util/src/journaldb/overlayrecentdb.rs

```diff
@@ -430,8 +434,51 @@ impl HashDB for OverlayRecentDB {
 		v.or_else(|| self.payload(key))
 	}
 
+	fn get_exec<'this>(
+		&'this self,
+		key: &'this H256,
+		f: &'this mut for<'a: 'this> FnMut(&'a [u8]),
+	) {
+		use ::std::cell::Cell;
+
+		let k = self.transaction_overlay.raw_ref(key);
+
+		if let Some((d, rc)) = k {
+			if rc > 0 {
+				f(d.as_ref());
+				return;
+			}
+		}
+
+		let was_called = Cell::new(false);
+
+		// This type annotation is necessary, the inferred lifetime is too restrictive otherwise.
+		// Possible bug in the Rust compiler.
+		let mut wrapper = |d: &[u8]| {
+			was_called.set(true);
+			f(d);
+		};
+
+		let journal_overlay = self.journal_overlay.read();
+		let short_key = to_short_key(key);
+
+		journal_overlay.backing_overlay.get_exec(&short_key, &mut wrapper);
```

Comment:
Damn, excellent point. I think a better solution might be to implement `get_with` for monomorphic code so that you get this speedup for free. Maybe I'll stick a TODO there, put in a performance issue and write the `get_ref` version.